crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +463,52 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if there is just a single inhabitant for this type.
+    ///
+    /// Note: This function aims to have no false positives, but might return `false`
+    /// for more complicated types that are actually singletons.
+    pub(crate) fn is_singleton(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Any
+            | Type::Never
+            | Type::Unknown
+            | Type::Todo
+            | Type::Unbound
+            | Type::Module(..)
+            | Type::Instance(..)
```

Comment:
This is fine for now, but note that some `Instance` types can also be singletons:
- Instances of `types.EllipsisType` (there is only one, `...`)
- Instances of `types.NotImplementedType` (there is only one, `NotImplemented`)

We could implement this using the `KnownClass` enum -- we could add an `is_singleton` method to that enum. This could be done as as followup PR

Reply:
Could possibly add a short TODO comment here so we don't forget about it

Reply:
> Turns out to be a bit harder than I expected, so I'm postponing it to another PR.

After trying to implement this, @AlexWaygood and me agreed to skip support for `EllipsisType`/`NotImplementedType` for now. The reason is that we can't look up the symbol types from `types`, since they are conditionally added only if the Python version is >=3.10: https://github.com/astral-sh/ruff/blob/74bf4b0653c0273af893fe81d31c126b45329ac6/crates/red_knot_vendored/vendor/typeshed/stdlib/types.pyi#L614-L620